### ISSUE TITLE
HA self-hosted cluster with test is failing

### ISSUE THREAD

**Issue by killianmuldoon** [kubernetes-sigs/cluster-api#9522]
Periodic CAPI e2e is failing. The test `When testing Cluster API working on self-hosted clusters using ClusterClass with a HA control plane [ClusterClass]` has failed for four consecutive runs. It's not directly linked to any PR merged in the CAPI repo AFAICT.

This test has been flaky recently so this might be four flakes in a row.

Test grid link: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main
Failing test run: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1709485228130570240

**Comment by killianmuldoon** [comment]
/triage accepted


@kubernetes-sigs/cluster-api-release-team

**Comment by nawazkh** [comment]
[This](https://kubernetes.slack.com/archives/C8TSNPY4T/p1696414810813339) is the thread I have been posting my analysis so far.

**Comment by killianmuldoon** [comment]
These tests fail with three different errors:
- `etcdImageTagCondition: expected 3 pods, got 4` https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*#aa41f168b84bfaa321bf
- `old nodes remain` https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*#1d058cd63b8638de3e75
- `error adding delete-for-move annotation` https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*#ac8b8ee04f7fcd030081

These failures only seem to happen on the main branch - we're looking for issues that could be responsible in the diff between these branches.

**Comment by killianmuldoon** [comment]
From the first few days after merging https://github.com/kubernetes-sigs/cluster-api/pull/9570 it looks like problem 1 and problem 2, i.e. `etcdImageTagCondition` and `old nodes remain` failures have been fixed. 

Toward the end of the week if we don't get any more of those failures we should consider them solved. IMO that should be enough to remove `release-blocking` from this issue, but we should consider that once we have more data in a few days.

The third failure `error adding delete-for-move annotation` is still occurring though, and we should look to fix it, though it's not the highest priority flake right now.

**Comment by nawazkh** [comment]
I see one flake with `old nodes remain` on `capi-mink8s-e2e-main`: reference run [1716485778235723776](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-mink8s-main/17164857782357237760). 

Moreover, there are almost consistent flakes on `When testing Cluster API working on self-hosted clusters using ClusterClass with a HA control plane [ClusterClass] Should pivot the bootstrap cluster to a self-hosted cluster` test. I am taking a look at the the diff between a green run and a flake. 

Are we sure we want to drop the "release-blocking" tag from this issue?

**Comment by killianmuldoon** [comment]
@nawazkh you're right - I think the state is still pretty unstable so we have to keep release-blocking for now. We have some additional `old nodes remain` failures as well as a cluster of `error adding delete-for-move annotation` failures.

Let's keep release blocking on this until those are in a better state.

**Comment by killianmuldoon** [comment]
Current state:
- `etcdImageTagCondition` is fixed by https://github.com/kubernetes-sigs/cluster-api/pull/9570
The other two errors are still occurring flakily.

**Comment by killianmuldoon** [comment]
After spending some time on it I've come up with a theory for what's causing `old nodes remain`. 


1) CAPI components are often deployed on the MachinePool instance in the self-hosted cluster.
2) It is the responsibility of the CAPI MachinePool controller to clean up the nodes of MachinePools when they are no longer listed in `Spec.ProviderIDList` 
3) When an upgrade happens the DockerMachinePool deletes the container running a MachinePool. This ungracefully stops the node and all running pods.
4) The capi-controller manager is dead at this point. It can not clean up the old nodes of the MachinePool. From a Kubernetes perspective the Node is `NotReady` but it doesn't clean up the pods.
5) This happens during the control plane upgrade. Should workers be upgraded at all while the control plane is still upgrading? 
6) Eventually something (probably the deployment controller, but it seems slower than the healthcheck would predict) redeploys the CAPI pods from the dead node to the cluster, but this process can take time. The controllers then take some time to start up causing this upgrade to take much longer.
7) Once the capi controllers are running again the MP nodes are eventually deleted and the rest of the cluster is eventually upgraded.

The additional time spent waiting for CAPI to recover from the ungraceful shutdown of the MachinePool is the root cause of this issue AFAICT.

I think this is a bug that can be fixed by finding a safer way to remove the docker containers that the MachinePools are running on. We could try to only delete the infrastructure after the Node has already been deleted.

We should consider removing MachinePools from this test after the RC next week to unblock the release. We can duplicate the jobs to keep some coverage so folks can work on fixing those tests.


I'm not sure if there's some explanation leading from this for the `error adding delete-for-move annotation` bug. I'm going to add some logging to try to figure out if something similar is going on.

@nawazkh @kubernetes-sigs/cluster-api-release-team 

@willie-yao (Seeing as you implemented the MP e2e testing): WDYT about removing MPs for this test for this release?

**Comment by vincepri** [comment]
-1 on removing tests, or disabling them. If we need, we'll delay the release indefinitely until either the tests are fixed, or the related code (not the tests) is reverted/removed.

**Comment by willie-yao** [comment]
> (Seeing as you implemented the MP e2e testing): WDYT about removing MPs for this test for this release?

I'm +1 on removing them temporarily. I don't see a connection between this issue and the addition of MachinePools to ClusterClass, rather MachinePools themselves. Since this issue has existed before the test is added, I think it'll be okay to remove them temporarily to unblock the release.

**Comment by willie-yao** [comment]
I'm +1 on removing this as a last resort but I think ideally we should be fixing this ahead of the release. Just to be clear: This is only happening on the ClusterClass self-hosted test? And is this because of ClusterClass, or the existence of MachinePools in general? I think the regular self-hosted test without ClusterClass doesn't include MachinePools. If the bug is due to only MachinePools and not ClusterClass, I'm more inclined to remove the test temporarily since it is a pre-existing bug

**Comment by CecileRobertMichon** [comment]
https://github.com/kubernetes-sigs/cluster-api/pull/8842 will fix this

**Comment by CecileRobertMichon** [comment]
This should now be fixed. Let's observe the CI signal until next Tuesday (11/21) to confirm we no longer see the flakes.

**Comment by furkatgofurov7** [comment]
> This should now be fixed. Let's observe the CI signal until next Tuesday (11/21) to confirm we no longer see the flakes.

Looks like it is green after the fix, but will leave it up to CI team Lead and members to confirm. 

cc @nawazkh

**Comment by adilGhaffarDev** [comment]
this is fixed on the `main`. I don't see any extraordinary flakes related to that.
I still see this flake: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1725069624287956992 , https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*#c6ca02bf53df8d27d1c7 
it happens rarely. We will look into it.

**Comment by furkatgofurov7** [comment]
> this is fixed on the `main`. I don't see any extraordinary flakes related to that. I still see this flake: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1725069624287956992 , https://storage.googleapis.com/k8s-triage/index.html?job=.*-cluster-api-.*&xjob=.*-provider-.*#c6ca02bf53df8d27d1c7 it happens rarely. We will look into it.

Thanks. Let's create a separate issue for the new flake (which is different than the ones described here) if needed and close the current one. Feel free to re-open in the future, if you encounter the same flakes. 

/close
